src/main/java/org/gradle/profiler/studio/StudioGradleClient.java

```diff
@@ -71,23 +82,48 @@ public BuildActionResult sync(List<String> gradleArgs, List<String> jvmArgs) {
+            Pair<StudioSyncRequestResult, BuildActionResult> pair = waitForSyncToFinish(connections, gradleArgs, jvmArgs);
+            StudioSyncRequestResult syncRequestResult = pair.getLeft();
+            BuildActionResult durationResult = pair.getRight();
+            System.out.printf("* Full Gradle execution time: %dms%n", durationResult.getGradleToolingAgentExecutionTime().get().toMillis());
+            System.out.printf("* Full IDE execution time: %dms%n", durationResult.getIdeExecutionTime().get().toMillis());
+            System.out.printf("* Full sync has completed in: %dms and it %s%n", durationResult.getExecutionTime().toMillis(), syncRequestResult);
+            return durationResult;
         });
     }
 
+    private Pair<StudioSyncRequestResult, BuildActionResult> waitForSyncToFinish(StudioConnections connections, List<String> gradleArgs, List<String> jvmArgs) {
+        System.out.println("* Sync has started, waiting for it to complete...");
+        AtomicBoolean isSyncRequestCompleted = new AtomicBoolean();
+        CompletableFuture<Long> gradleInvocations = CompletableFuture.supplyAsync(() -> collectGradleInvocations(connections, isSyncRequestCompleted, gradleArgs, jvmArgs), executor);
+        StudioSyncRequestCompleted syncRequestCompleted = connections.getPluginConnection().receiveSyncRequestCompleted(SYNC_REQUEST_COMPLETED_TIMEOUT);
+        isSyncRequestCompleted.set(true);
+        long gradleInvocationsDuration = gradleInvocations.join();
+        BuildActionResult result = BuildActionResult.withIdeTimings(
+            Duration.ofMillis(syncRequestCompleted.getDurationMillis()),
+            Duration.ofMillis(gradleInvocationsDuration),
+            Duration.ofMillis(syncRequestCompleted.getDurationMillis() - gradleInvocationsDuration)
+        );
+        return Pair.of(syncRequestCompleted.getResult(), result);
+    }
+
+    private long collectGradleInvocations(StudioConnections connections, AtomicBoolean isSyncRequestCompleted, List<String> gradleArgs, List<String> jvmArgs) {
+        long durations = 0;
+        int invocation = 1;
+        ServerConnection agentConnection = connections.getAgentConnection();
+        while (!isSyncRequestCompleted.get()) {
+            Optional<GradleInvocationStarted> invocationStarted = agentConnection.maybeReceiveGradleInvocationStarted(GRADLE_INVOCATION_STARTED_TIMEOUT);
+            if (invocationStarted.isPresent()) {
+                System.out.printf("* Gradle invocation %s has started, waiting for it to complete...%n", invocation);
+                agentConnection.send(new GradleInvocationParameters(gradleArgs, jvmArgs));
+                GradleInvocationCompleted agentCompleted = agentConnection.receiveGradleInvocationCompleted(GRADLE_INVOCATION_COMPLETED_TIMEOUT);
+                System.out.printf("* Gradle invocation %s has completed in: %sms%n", invocation++, agentCompleted.getDurationMillis());
+                durations += agentCompleted.getDurationMillis();
```

Comment:
Would it make sense to collect the individual invocations here and put them in the report?

Reply:
Good point, I think it makes sense yeah.